File: HW6/Driver.c
/**
 * Name: ChakMan SIo
 * Class: CSE 12
 * Account cs12xef
 * Quarter: Fall 2016
 * Assignment: HW6 
 * File name: Driver.c
 * Class Description: This class allow user to acces all 
 *                    the ucsd student, we can use those
 *                    operator in Hash.c
 *              
 */

#include <string.h>
#include <stdio.h>
#include <iostream>
#include <getopt.h>
#include "Base.h"
#include "SymTab.h"
#include <cstdlib>
using namespace std;

#ifdef NULL
#undef NULL
#define NULL 0
#endif

class UCSDStudent : public Base {
	char * name;
	long studentnum;
public:
	UCSDStudent (char * nm, long sn) :
  name (strdup (nm)), studentnum (sn) {}
	~UCSDStudent (void) {
    free (name);
	}
  // get name in c++
  operator char * (void) {  
    return name;  
  }

  operator long (void) ;  
	// operator is equal
  long operator == (Base & base)  {  
    return ! strcmp (name, base);  
  } 
  // operator is less than
  long operator < (Base & base) {
		return (strcmp (name, base) < 0) ? 1 : 0;
  }

  ostream & Write (ostream & stream) const {
		return stream << "name:  " << name
			<< "  Studentnum:  " << studentnum;
	}
};

/* You may have one or more functions defined here, too */
UCSDStudent :: operator long(void) {
  // it is for calculate ascii sum
  long sum = 0;  
    for (long index = 0; name[index]; index++) 
      sum += name[index];
      return sum; 
}


int main (int argc, char * const * argv) {
	SymTab ST (5);
	char buffer[128];
	char command;
	long number;
	char option;

	/* initialize debug variable */
	HashTable::Set_Debug(0);

	 /* check command line options for debug display */
	 while ((option = getopt (argc, argv, "x")) != EOF) {

		 switch (option) {
			 case 'x': HashTable::Set_Debug(1);
				 break;
		 	}
	} 

	ST.Write (cout << "Initial Symbol Table:\n");
	while (cin) {
		command = NULL;		// reset command each time in loop
		cout << "Please enter a command:  ((i)nsert, (l)ookup, (w)rite):  ";
		cin >> command;

		switch (command) {
		case 'i':
			cout << "Please enter UCSD student name to insert:  ";
			cin >> buffer;	// formatted input

			cout << "Please enter UCSD student number:  ";
			cin >> number;

			// create student and place in symbol table
			if(!ST.Insert (new UCSDStudent (buffer, number))){

				ST.Write (cout << "\nFinal Symbol Table:\n");
				exit (0);
			}
			break;

		case 'l': {
			const Base * found;	// whether found or not

			cout << "Please enter UCSD student name to lookup:  ";
			cin >> buffer;	// formatted input

			UCSDStudent stu (buffer, 0);
			found = ST.Lookup (& stu);
			
			if (found)
				found->Write (cout << "Student found!!!\n") << "\n";
			else
				cout << "Student " << buffer << " not there!\n";
			}
			break;

		case 'w':
			ST.Write (cout << "The Symbol Table contains:\n");
		}
	}
	ST.Write (cout << "\nFinal Symbol Table:\n");
}

File: HW6/Driver.java
/**
 * Name: ChakMan Sio 
 * Class: CSE 12 
 * Account cs12xef
 * Quarter: Fall 2016 
 * Assignment: HW6 
 * File name: Driver.java 
 * Class Description: this class allow to accsss the UCSDStudent, I can 
 *                    use those functions to implement my HashTable.java
 */

import java.io.*;

class UCSDStudent extends Base {
        private String name;
        private long studentnum;
        
				public long value;
        public UCSDStudent (String nm, long val){
            name = new String (nm);
						value = val;
				}
				public UCSDStudent (Variable variable){
            name = new String (variable.name);
            value = variable.value;

				}

				public String getName(){
            return name;
        }

				public boolean equals (Object object){
            Base bbb = (Base) object;
				     return name.equals (bbb.getName ());
        }
				public boolean isLessThan (Base bbb) {
				    return (name.compareTo (bbb.getName ()) < 0) ? true : false;
				}
				public String toString () {
				    return "name: "+ name +" "+ "Studentnum: " + value ;
						                                                     
        }

				public int hashCode () {
				    int retval = 0;
            int index = 0;

				    while (index != name.length ()) {
				        retval += name.charAt (index);
								index ++;
				    }

				    return retval;
        }
}

public class Driver {
        private static final short NULL = 0;

        public static void main (String [] args) {
                /* initialize debug states */
                HashTable.setDebugOff();

                /* check command line options */
                for (int index = 0; index < args.length; ++index) {
                        if (args[index].equals("-x"))
                                HashTable.setDebugOn();
                }

                /* The real start of the code */
                SymTab symtab = new SymTab (5);
                String buffer = null;
                char command;
                long number = 0;

                System.out.print ("Initial Symbol Table:\n" + symtab);

                while (true) {
                        command = 0;    // reset command each time in loop
                        System.out.print ("Please enter a command:  "
                                + "((i)nsert, (l)ookup, (w)rite):  ");

                        try {
                        command = MyLib.getchar ();
                        MyLib.clrbuf (command); // get rid of return

                        switch (command) {
                        case 'i':
                                System.out.print (
                                "Please enter UCSD Student name to insert:  ");
                                buffer = MyLib.getline ();// formatted input

                                System.out.print (
                                        "Please enter UCSD Student number:  ");

                                number = MyLib.decin ();
                                MyLib.clrbuf (command); // get rid of return

                                // create Student and place in symbol table
                                if(!symtab.insert (
                                        new UCSDStudent (buffer, number))){

                                        System.out.print ("\nFinal Symbol "
                                                        + "Table:\n" + symtab);
                                        System.exit(0);
                                }
                                break;

                        case 'l': {
                                Base found;     // whether found or not

                                System.out.print (
                                "Please enter UCSD Student name to lookup:  ");

                                buffer = MyLib.getline ();// formatted input

                                UCSDStudent stu = new UCSDStudent (buffer, 0);
                                found = symtab.lookup (stu);
                                
                                if (found != null) {
                                        System.out.println ("Student found!!!");
                                        System.out.println (found);
                                }
                                else
                                        System.out.println ("Student " + buffer
                                                + " not there!");
                                }
                                break;

                        case 'w':
                                System.out.print (
                                    "The Symbol Table contains:\n" + symtab);
                        }
                        }
                        catch (EOFException eof) {
                                break;
                        }
                }

                System.out.print ("\nFinal Symbol Table:\n" + symtab);
        }
}
